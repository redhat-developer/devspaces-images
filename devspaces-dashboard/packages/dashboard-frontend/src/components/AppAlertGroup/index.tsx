/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Alert, AlertActionCloseButton, AlertGroup, AlertVariant } from '@patternfly/react-core';
import React from 'react';
import { lazyInject } from '../../inversify.config';
import { AppAlerts } from '../../services/alerts/appAlerts';
import { AlertItem } from '../../services/helpers/types';

type Props = unknown;

type State = {
  alerts: AlertItem[];
};

class AppAlertGroup extends React.PureComponent<Props, State> {
  @lazyInject(AppAlerts)
  private readonly appAlerts: AppAlerts;

  private readonly showAlertHandler: (alerts: AlertItem[]) => void;

  constructor(props: Props) {
    super(props);

    this.state = {
      alerts: [],
    };
    this.showAlertHandler = (alerts: AlertItem[]) => {
      this.setState({ alerts });
    };
  }

  public componentDidMount(): void {
    this.appAlerts.subscribe(this.showAlertHandler);
  }

  public componentWillUnmount(): void {
    this.appAlerts.unsubscribe(this.showAlertHandler);
  }

  private getTime(variant: AlertVariant): number {
    let time: number;

    switch (variant) {
      case AlertVariant.success:
        time = 2000;
        break;
      case AlertVariant.info:
        time = 8000;
        break;
      default:
        time = 20000;
    }

    return time;
  }

  private getAlert(item: AlertItem): React.ReactElement {
    const { variant, title, key, children } = item;
    const showAlertTimer = setTimeout(() => {
      this.appAlerts.removeAlert(key);
    }, this.getTime(variant));
    return (
      <Alert
        variant={variant}
        title={title}
        key={key}
        actionClose={
          <AlertActionCloseButton
            onClose={() => {
              clearTimeout(showAlertTimer);
              this.appAlerts.removeAlert(key);
            }}
          />
        }
      >
        {children ? children : ''}
      </Alert>
    );
  }

  public render(): React.ReactElement {
    return <AlertGroup isToast>{this.state.alerts.map(alert => this.getAlert(alert))}</AlertGroup>;
  }
}

export default AppAlertGroup;
