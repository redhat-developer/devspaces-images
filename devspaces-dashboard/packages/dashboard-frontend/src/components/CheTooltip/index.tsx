/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Tooltip, TooltipProps } from '@patternfly/react-core';
import React from 'react';

export type Props = Omit<TooltipProps, 'ref'>;

export class CheTooltip extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    return (
      <Tooltip
        isContentLeftAligned={true}
        style={{ border: '1px solid', borderRadius: '3px', opacity: '0.9' }}
        {...this.props}
      />
    );
  }
}
