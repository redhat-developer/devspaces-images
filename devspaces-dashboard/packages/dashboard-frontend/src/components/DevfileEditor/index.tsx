/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { connect, ConnectedProps } from 'react-redux';
import { AppState } from '../../store';
import { DisposableCollection } from '../../services/helpers/disposable';
import {
  ProtocolToMonacoConverter,
  MonacoToProtocolConverter,
} from 'monaco-languageclient/lib/monaco-converter';
import { languages, editor, Range, Position, IRange } from 'monaco-editor-core';
import {
  TextDocument,
  getLanguageService,
  LanguageService,
  CompletionItem,
} from 'yaml-language-server';
import { initDefaultEditorTheme } from '../../services/monacoThemeRegister';
import stringify, { language, conf } from '../../services/helpers/editor';
import $ from 'jquery';
import { merge, isMatch } from 'lodash';
import devfileApi from '../../services/devfileApi';
import { selectDevfileSchema } from '../../store/DevfileRegistries/selectors';
import { selectPlugins } from '../../store/Plugins/chePlugins/selectors';
import { selectBranding } from '../../store/Branding/selectors';

import './DevfileEditor.styl';

const LANGUAGE_ID = 'yaml';
const YAML_SERVICE = 'yamlService';
const MONACO_CONFIG: editor.IStandaloneEditorConstructionOptions = {
  language: 'yaml',
  wordWrap: 'on',
  lineNumbers: 'on',
  scrollBeyondLastLine: false,
  readOnly: false,
};

(self as any).MonacoEnvironment = {
  getWorkerUrl: () => './editor.worker.js',
};

type Props = MappedProps & {
  devfile: che.WorkspaceDevfile | devfileApi.Devfile;
  decorationPattern?: string;
  onChange: (newValue: string, isValid: boolean) => void;
  isReadonly?: boolean;
  additionSchema?: { [key: string]: any };
};
type State = {
  errorMessage: string;
};

export class DevfileEditor extends React.PureComponent<Props, State> {
  public static EDITOR_THEME: string | undefined;
  private readonly toDispose = new DisposableCollection();
  private handleResize: () => void;
  private editor: editor.IStandaloneCodeEditor;
  private readonly yamlService: LanguageService;
  private m2p = new MonacoToProtocolConverter();
  private p2m = new ProtocolToMonacoConverter();
  private createDocument = (model): TextDocument =>
    TextDocument.create(
      'inmemory://model.yaml',
      model.getModeId(),
      model.getVersionId(),
      model.getValue(),
    );
  private skipNextOnChange: boolean;

  constructor(props: Props) {
    super(props);

    this.state = {
      errorMessage: '',
    };

    // lazy initialization
    if (!window[YAML_SERVICE]) {
      this.yamlService = getLanguageService(() => Promise.resolve(''), {
        resolveRelativePath: () => '',
      });
      window[YAML_SERVICE] = this.yamlService;
    } else {
      this.yamlService = window[YAML_SERVICE];
      this.updateSchema();
      return;
    }
    if (!DevfileEditor.EDITOR_THEME) {
      // define the default
      DevfileEditor.EDITOR_THEME = initDefaultEditorTheme();
    }
    if (
      languages?.getLanguages &&
      languages.getLanguages().find(language => language.id === LANGUAGE_ID) === undefined
    ) {
      // register the YAML language with Monaco
      languages.register({
        id: LANGUAGE_ID,
        extensions: ['.yaml', '.yml'],
        aliases: ['YAML', 'yaml'],
        mimetypes: ['application/yaml'],
      });
      languages.setMonarchTokensProvider(LANGUAGE_ID, language);
      languages.setLanguageConfiguration(LANGUAGE_ID, conf);
    }
    // register language server providers
    this.registerLanguageServerProviders();
    this.updateSchema();
  }

  private updateSchema(): void {
    let jsonSchema = this.props.devfileSchema ? Object.assign({}, this.props.devfileSchema) : {};
    const additionSchema = this.props.additionSchema;
    if (additionSchema) {
      if (jsonSchema?.oneOf) {
        jsonSchema.oneOf = jsonSchema.oneOf.map(schema => merge({}, schema, additionSchema));
      } else {
        jsonSchema = merge({}, jsonSchema, additionSchema);
      }
    }
    const schemas = [{ uri: 'inmemory:yaml', fileMatch: ['*'], schema: jsonSchema }];
    this.yamlService.configure({ validate: true, hover: true, schemas, completion: true });
  }

  public updateContent(devfile: che.WorkspaceDevfile | devfileApi.Devfile): void {
    if (!this.editor) {
      return;
    }
    this.skipNextOnChange = true;
    const doc = this.editor.getModel();
    doc?.setValue(stringify(devfile));
  }

  public componentDidUpdate(prevProps: Props): void {
    if (this.handleResize) {
      this.handleResize();
    }
    if (!isMatch(prevProps.additionSchema || {}, this.props.additionSchema || {})) {
      this.updateSchema();
    }
  }

  // This method is called when the component is first added to the document
  public componentDidMount(): void {
    const element = $('.devfile-editor .monaco').get(0);
    if (element) {
      const value = stringify(this.props.devfile);
      this.editor = editor.create(
        element,
        Object.assign(
          { value },
          {
            ...MONACO_CONFIG,
            readOnly: !!this.props.isReadonly,
          },
        ),
      );
      const doc = this.editor.getModel();
      doc?.updateOptions({ tabSize: 2 });

      const handleResize = (): void => {
        const layout = { height: element.offsetHeight, width: element.offsetWidth };
        // if the element is hidden
        if (layout.height === 0 || layout.width === 0) {
          return;
        }
        this.editor.layout(layout);
      };
      this.handleResize = handleResize;
      window.addEventListener('resize', handleResize);
      this.toDispose.push({
        dispose: () => {
          window.removeEventListener('resize', handleResize);
        },
      });

      let oldDecorationIds: string[] = []; // Array containing previous decorations identifiers.
      const updateDecorations = (): void => {
        if (this.props.decorationPattern) {
          oldDecorationIds = this.editor.deltaDecorations(oldDecorationIds, this.getDecorations());
        }
      };
      updateDecorations();
      doc?.onDidChangeContent(() => {
        updateDecorations();
        this.onChange(this.editor.getValue(), true);
      });
      // init language server validation
      this.initLanguageServerValidation(this.editor);
    }

    const handleMessage = (event: MessageEvent): void => {
      if (typeof event.data !== 'string') {
        return;
      }
      const { data } = event;
      if (
        (data === 'show-navbar' || data === 'hide-navbar' || data === 'toggle-navbar') &&
        this.handleResize
      ) {
        this.handleResize();
      }
    };
    window.addEventListener('message', handleMessage, false);
    this.toDispose.push({
      dispose: () => {
        window.removeEventListener('message', handleMessage, false);
      },
    });
  }

  // This method is called when the component is removed from the document
  public componentWillUnmount(): void {
    this.toDispose.dispose();
  }

  public render(): React.ReactElement {
    const href = this.props.branding.docs.devfile;
    const { errorMessage } = this.state;

    return (
      <div className="devfile-editor">
        <div className="monaco">&nbsp;</div>
        <div className="error">{errorMessage}</div>
        <a target="_blank" rel="noopener noreferrer" href={href}>
          Devfile Documentation
        </a>
      </div>
    );
  }

  private getDecorations(): editor.IModelDecoration[] {
    const decorations: editor.IModelDecoration[] = [];
    if (this.props.decorationPattern) {
      const decorationRegExp = new RegExp(this.props.decorationPattern, 'img');
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const model = this.editor.getModel()!;
      const value = this.editor.getValue();
      let match = decorationRegExp.exec(value);
      while (match) {
        const startPosition = model.getPositionAt(match.index);
        const endPosition = model.getPositionAt(match.index + match[0].length);
        const range = new Range(
          startPosition.lineNumber,
          startPosition.column,
          endPosition.lineNumber,
          endPosition.column,
        );
        const options: editor.IModelDecorationOptions = {
          inlineClassName: 'devfile-editor-decoration',
        };
        decorations.push({
          range,
          options,
        } as editor.IModelDecoration);
        match = decorationRegExp.exec(value);
      }
    }
    return decorations;
  }

  private onChange(newValue: string, isValid: boolean): void {
    if (this.skipNextOnChange) {
      this.skipNextOnChange = false;
      return;
    }
    this.props.onChange(newValue, isValid);
  }

  private registerLanguageServerProviders(): void {
    const createDocument = this.createDocument;
    const yamlService = this.yamlService;
    const m2p = this.m2p;
    const p2m = this.p2m;

    const sortTextCache = {};
    const createSortText = (pluginId: string): string => {
      if (sortTextCache[pluginId]) {
        return sortTextCache[pluginId];
      }
      const [publisher, name, version] = pluginId.split('/');
      const semverRE = /^\d+?\.\d+?\.\d+?/;
      let versionText: string;
      if (semverRE.test(version)) {
        let [major, minor, rest] = version.split('.');
        major = addLeadingZeros(major);
        minor = addLeadingZeros(minor);
        rest = rest.replace(/^(\d+)/, addLeadingZeros);
        versionText = `${major}.${minor}.${rest}`;
      } else {
        let prefix = '-';
        if (version === 'latest') {
          prefix += '0';
        } else if (version === 'next') {
          prefix += '1';
        } else if (version === 'nightly') {
          prefix += '2';
        }
        versionText = `${prefix}${version}`;
      }
      const sortText = `${publisher}/${name}/${versionText}`;
      sortTextCache[pluginId] = sortText;
      return sortText;
    };
    const addLeadingZeros = (number: string): string => {
      while (number.length < 4) {
        number = '0' + number;
      }
      return number;
    };

    languages.registerCompletionItemProvider(LANGUAGE_ID, {
      provideCompletionItems(model: editor.ITextModel, position: Position) {
        const document = createDocument(model);
        return yamlService
          .doComplete(document, m2p.asPosition(position.lineNumber, position.column), true)
          .then(list => {
            const completionResult = p2m.asCompletionResult(
              list as any,
              {
                startColumn: position.column,
                startLineNumber: position.lineNumber,
                endColumn: position.column,
                endLineNumber: position.lineNumber,
              } as IRange,
            );
            if (!completionResult || !completionResult.suggestions) {
              return completionResult;
            }
            // convert completionResult into suggestions
            const defaultInsertTextRules = languages.CompletionItemInsertTextRule.InsertAsSnippet;
            const suggestions = completionResult.suggestions.map(suggestion => {
              return Object.assign(suggestion, {
                insertTextRules: suggestion.insertTextRules
                  ? suggestion.insertTextRules
                  : defaultInsertTextRules,
                sortText: createSortText(suggestion.insertText),
              });
            });
            return { suggestions };
          });
      },
      async resolveCompletionItem(model, range, item) {
        return (yamlService as any)
          .doResolve(m2p.asCompletionItem(item))
          .then((result: CompletionItem) => p2m.asCompletionItem(result as any, range));
      },
    } as any);
    languages.registerDocumentSymbolProvider(LANGUAGE_ID, {
      provideDocumentSymbols(model: editor.ITextModel) {
        return p2m.asSymbolInformations(yamlService.findDocumentSymbols(createDocument(model)));
      },
    });
    languages.registerHoverProvider(LANGUAGE_ID, {
      async provideHover(model: editor.ITextModel, position: Position) {
        const hover = await yamlService.doHover(
          createDocument(model),
          m2p.asPosition(position.lineNumber, position.column),
        );
        return p2m.asHover(hover);
      },
    });
  }

  private initLanguageServerValidation(_editor: editor.IStandaloneCodeEditor): void {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const model = _editor.getModel()!;
    let validationTimer: number;

    model.onDidChangeContent(() => {
      const document = this.createDocument(model);
      this.setState({ errorMessage: '' });
      if (validationTimer) {
        clearTimeout(validationTimer);
      }
      validationTimer = window.setTimeout(() => {
        this.yamlService.doValidation(document, false).then(diagnostics => {
          const markers = this.p2m.asDiagnostics(diagnostics) as any;
          let errorMessage = '';
          if (Array.isArray(markers) && markers.length > 0) {
            const { message, startLineNumber, startColumn } = markers[0];
            if (startLineNumber && startColumn) {
              errorMessage += `line[${startLineNumber}] column[${startColumn}]: `;
            }
            errorMessage += `Error. ${message}`;
          }
          if (errorMessage) {
            this.setState({ errorMessage: `Error. ${errorMessage}` });
            this.onChange(_editor.getValue(), false);
          }
          editor.setModelMarkers(model, 'default', markers ? markers : []);
        });
      });
    });
  }
}

const mapStateToProps = (state: AppState) => ({
  branding: selectBranding(state),
  devfileSchema: selectDevfileSchema(state),
  plugins: selectPlugins(state),
});

const connector = connect(mapStateToProps, null, null, { forwardRef: true });

type MappedProps = ConnectedProps<typeof connector>;
export default connector(DevfileEditor);
