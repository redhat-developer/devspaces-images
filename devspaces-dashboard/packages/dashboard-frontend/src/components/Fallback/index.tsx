/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Spinner, spinnerSize } from '@patternfly/react-core';
import React from 'react';

const Fallback = (
  <div style={{ height: '100%', textAlign: 'center', opacity: '0.5' }}>
    <Spinner
      data-testid="fallback-spinner"
      size={spinnerSize.xl}
      style={{ top: 'calc(50% - 18px)' }}
    />
  </div>
);

export default Fallback;
