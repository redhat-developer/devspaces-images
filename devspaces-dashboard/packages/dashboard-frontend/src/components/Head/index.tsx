/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { Helmet } from 'react-helmet';
import { connect, ConnectedProps } from 'react-redux';
import { AppState } from '../../store';
import { selectBranding } from '../../store/Branding/selectors';

type Props = MappedProps & {
  pageName?: string;
};

class Head extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    const {
      pageName,
      branding: { title },
    } = this.props;

    return (
      <Helmet>
        <title>{pageName ? `${title} | ${pageName}` : title}</title>
      </Helmet>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  branding: selectBranding(state),
});

const connector = connect(mapStateToProps);

type MappedProps = ConnectedProps<typeof connector>;
export default connector(Head);
