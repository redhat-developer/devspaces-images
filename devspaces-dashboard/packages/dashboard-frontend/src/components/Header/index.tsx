/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import {
  Flex,
  FlexItem,
  PageSection,
  PageSectionVariants,
  Text,
  TextContent,
  TextVariants,
} from '@patternfly/react-core';
import WorkspaceStatusLabel from '../WorkspaceStatusLabel';
import {
  DeprecatedWorkspaceStatus,
  DevWorkspaceStatus,
  WorkspaceStatus,
} from '../../services/helpers/types';

const SECTION_THEME = PageSectionVariants.light;

type Props = {
  status: WorkspaceStatus | DevWorkspaceStatus | DeprecatedWorkspaceStatus;
  title: string;
};

class Header extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    const { title, status } = this.props;

    return (
      <PageSection variant={SECTION_THEME}>
        <Flex>
          <FlexItem alignSelf={{ default: 'alignSelfCenter' }}>
            <TextContent>
              <Text component={TextVariants.h1}>{title}</Text>
            </TextContent>
          </FlexItem>
          <FlexItem>
            <WorkspaceStatusLabel status={status} />
          </FlexItem>
        </Flex>
      </PageSection>
    );
  }
}

export default Header;
