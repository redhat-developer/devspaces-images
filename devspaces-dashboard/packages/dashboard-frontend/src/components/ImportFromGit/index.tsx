/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  Button,
  ButtonVariant,
  Flex,
  FlexItem,
  Form,
  FormGroup,
  FormHelperText,
  Panel,
  PanelHeader,
  PanelMain,
  PanelMainBody,
  TextInput,
  Title,
  ValidatedOptions,
} from '@patternfly/react-core';
import { ExclamationCircleIcon } from '@patternfly/react-icons';
import { History } from 'history';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import { validateLocation } from '@/components/ImportFromGit/helpers';
import RepoOptionsAccordion from '@/components/ImportFromGit/RepoOptionsAccordion';
import UntrustedSourceModal from '@/components/UntrustedSourceModal';
import { FactoryLocationAdapter } from '@/services/factory-location-adapter';
import { EDITOR_ATTR, EDITOR_IMAGE_ATTR } from '@/services/helpers/factoryFlow/buildFactoryParams';
import { buildUserPreferencesLocation } from '@/services/helpers/location';
import { UserPreferencesTab } from '@/services/helpers/types';
import { AppState } from '@/store';
import { selectSshKeys } from '@/store/SshKeys/selectors';
import * as WorkspacesStore from '@/store/Workspaces';

const FIELD_ID = 'git-repo-url';

export type Props = MappedProps & {
  history: History;
  editorDefinition: string | undefined;
  editorImage: string | undefined;
};
export type State = {
  hasSshKeys: boolean;
  location: string;
  locationValidated: ValidatedOptions;
  remotesValidated: ValidatedOptions;
  isFocused: boolean;
  isConfirmationOpen: boolean;
};

class ImportFromGit extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      hasSshKeys: this.props.sshKeys.length > 0,
      locationValidated: ValidatedOptions.default,
      location: '',
      remotesValidated: ValidatedOptions.default,
      isFocused: false,
      isConfirmationOpen: false,
    };
  }

  public componentDidUpdate(prevProps: Readonly<Props>, prevState: Readonly<State>): void {
    const { location, isFocused } = this.state;
    if (!isFocused && (location !== prevState.location || prevState.isFocused)) {
      const inputElement = document.getElementById(FIELD_ID) as HTMLInputElement;
      if (inputElement) {
        inputElement.value = location;
      }
    }
  }

  private openConfirmationDialog(): void {
    this.setState({ isConfirmationOpen: true });
  }

  private handleConfirmationOnClose(): void {
    this.setState({ isConfirmationOpen: false });
  }

  private handleConfirmationOnContinue(): void {
    this.setState({ isConfirmationOpen: false });
    this.startFactory();
  }

  private handleCreate(): void {
    this.openConfirmationDialog();
  }

  private startFactory(): void {
    const { editorDefinition, editorImage } = this.props;
    const factory = new FactoryLocationAdapter(this.state.location);

    // add the editor definition and editor image to the URL
    // if they are not already there
    if (!factory.searchParams.has(EDITOR_ATTR) && !factory.searchParams.has(EDITOR_IMAGE_ATTR)) {
      if (editorDefinition !== undefined) {
        factory.searchParams.set(EDITOR_ATTR, editorDefinition);
      }
      if (editorImage !== undefined) {
        factory.searchParams.set(EDITOR_IMAGE_ATTR, editorImage);
      }
    }

    // open a new page to handle that
    window.open(`${window.location.origin}/#${factory.toString()}`, '_blank');
  }

  private handleChange(location: string): void {
    location = location.trim();
    if (this.state.location === location) {
      return;
    }
    const validated = validateLocation(location, this.state.hasSshKeys);
    this.setState({ locationValidated: validated, location });
  }

  private getErrorMessage(location: string): string | React.ReactNode {
    const isValidGitSsh = FactoryLocationAdapter.isSshLocation(location);

    if (isValidGitSsh && !this.state.hasSshKeys) {
      return (
        <FormHelperText icon={<ExclamationCircleIcon />} isHidden={false} isError={true}>
          No SSH keys found. Please add your SSH keys in the{' '}
          <Button variant="link" isInline onClick={() => this.openUserPreferences()}>
            User Preferences
          </Button>{' '}
          and then try again.
        </FormHelperText>
      );
    }

    return 'The URL or SSHLocation is not valid.';
  }

  private openUserPreferences(): void {
    const location = buildUserPreferencesLocation(UserPreferencesTab.SSH_KEYS);
    this.props.history.push(location);
  }

  public buildForm(): React.JSX.Element {
    const { location } = this.state;
    const { locationValidated, remotesValidated } = this.state;

    const buttonDisabled =
      !location ||
      locationValidated === ValidatedOptions.error ||
      remotesValidated === ValidatedOptions.error;
    const errorMessage = this.getErrorMessage(location);
    const validated =
      remotesValidated === ValidatedOptions.error ? ValidatedOptions.error : locationValidated;

    return (
      <Form
        isHorizontal={true}
        onSubmit={e => {
          e.preventDefault();
          if (buttonDisabled) {
            return false;
          }
          this.handleCreate();
        }}
      >
        <FormGroup
          fieldId={FIELD_ID}
          label="Git repo URL"
          isRequired={true}
          validated={validated}
          helperTextInvalid={errorMessage}
          helperTextInvalidIcon={<ExclamationCircleIcon />}
          helperText="Import from a Git repository to create your first workspace."
        >
          <Flex>
            <FlexItem grow={{ default: 'grow' }} style={{ maxWidth: '500px', minWidth: '70%' }}>
              <TextInput
                id={FIELD_ID}
                aria-label="HTTPS or SSH URL"
                placeholder="Enter HTTPS or SSH URL"
                validated={validated}
                onFocus={() => this.setState({ isFocused: true })}
                onBlur={() => this.setState({ isFocused: false })}
                onChange={value => this.handleChange(value)}
              />
            </FlexItem>
            <FlexItem>
              <Button
                id="create-and-open-button"
                isDisabled={buttonDisabled}
                variant={ButtonVariant.secondary}
                onClick={() => this.handleCreate()}
              >
                Create & Open
              </Button>
            </FlexItem>
          </Flex>
        </FormGroup>
      </Form>
    );
  }

  public render() {
    const { history } = this.props;
    const { isConfirmationOpen, location, locationValidated } = this.state;
    return (
      <>
        <UntrustedSourceModal
          location={location}
          isOpen={isConfirmationOpen}
          onContinue={() => this.handleConfirmationOnContinue()}
          onClose={() => this.handleConfirmationOnClose()}
        />
        <Panel>
          <PanelHeader>
            <Title headingLevel="h3">Import from Git</Title>
          </PanelHeader>
          <PanelMain>
            <PanelMainBody>{this.buildForm()}</PanelMainBody>
          </PanelMain>
          {locationValidated === ValidatedOptions.success && (
            <PanelMain>
              <PanelMainBody>
                <RepoOptionsAccordion
                  location={location}
                  history={history}
                  onChange={(location: string, remotesValidated: ValidatedOptions) => {
                    const locationValidated = validateLocation(location, this.state.hasSshKeys);
                    this.setState({ location, remotesValidated, locationValidated });
                  }}
                />
              </PanelMainBody>
            </PanelMain>
          )}
        </Panel>
      </>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  sshKeys: selectSshKeys(state),
});

const connector = connect(mapStateToProps, WorkspacesStore.actionCreators);

type MappedProps = ConnectedProps<typeof connector>;
export default connector(ImportFromGit);
