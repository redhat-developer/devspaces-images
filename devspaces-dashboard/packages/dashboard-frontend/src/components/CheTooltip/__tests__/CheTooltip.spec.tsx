/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { TooltipPosition } from '@patternfly/react-core';
import React from 'react';

import { CheTooltip, Props } from '@/components/CheTooltip';
import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';

// use actual CheTooltip component
jest.unmock('@/components/CheTooltip');

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

describe('CheTooltip component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot', () => {
    const props = {
      content: <span>Tooltip text.</span>,
    };

    const snapshot = createSnapshot(props);

    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('passed props', () => {
    const props: Props = {
      position: TooltipPosition.right,
      exitDelay: 500,
      content: <span>Tooltip text.</span>,
    };

    renderComponent(props);

    const tooltipProps = screen.getByTestId('tooltip-props');
    expect(tooltipProps).toHaveTextContent(
      '{"isContentLeftAligned":true,"style":{"border":"1px solid","borderRadius":"3px","opacity":"0.9"},"position":"right","exitDelay":500}',
    );

    const tooltipContent = screen.getByTestId('tooltip-content');
    expect(tooltipContent).toHaveTextContent('Tooltip text.');

    const tooltipPlacedTo = screen.getByTestId('tooltip-placed-to');
    expect(tooltipPlacedTo).toHaveTextContent('some text');

    screen.debug();
  });
});

function getComponent(props: Props): React.ReactElement {
  return (
    <CheTooltip {...props}>
      <div>some text</div>
    </CheTooltip>
  );
}
