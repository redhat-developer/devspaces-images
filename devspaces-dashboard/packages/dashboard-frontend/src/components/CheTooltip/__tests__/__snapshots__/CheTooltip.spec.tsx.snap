// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheTooltip component snapshot 1`] = `
<div
  data-testid="patternfly-tooltip"
>
  <span
    data-testid="tooltip-props"
  >
    {"isContentLeftAligned":true,"style":{"border":"1px solid","borderRadius":"3px","opacity":"0.9"}}
  </span>
  <div
    data-testid="tooltip-content"
  >
    <span>
      Tooltip text.
    </span>
  </div>
  <div
    data-testid="tooltip-placed-to"
  >
    <div>
      some text
    </div>
  </div>
</div>
`;
